migrate cogs to sqlalchemy

Jan200101 opened the pull request:
- adds a session attribute to the bot/client which can be used to start a database sesssion
- converts all uses of aioslite in cogs (and miscutils) to use sessions
- add missing database elements

I only tested this with some commands but it looks to function correctly.

Can you verify that the models match up with the database as it is on the server?
The next thing I'll add is proper migrations, so it would be great if the tables are correct.

Bobbyperson commented on pyproject.toml:
```diff
@@ -7,8 +7,7 @@ requires-python = ">=3.11,<3.13,!=3.12.5"
 dependencies = [
-    "aiosqlite>=0.21.0",
-    "sqlalchemy[asyncio]>=2.0.40",
+    "sqlalchemy[aiosqlite]>=2.0.40",
```
Did you mean `asyncio` here instead of `aiosqlite`?

Jan200101 replied:
`aiosqlite` is the same as `asyncio` but has an additional dependency on aiosqlite.
Since we use that as a driver, this is better.

Bobbyperson commented on cogs/economy.py:
```diff
@@ -1687,34 +1655,29 @@ async def sql(self, ctx, command: str | None = None):
-                    await cursor.execute(command)
-                    result = await cursor.fetchall()
+            async with self.client.session as session:
+                result = await session.execute(text(command)).fetchall()
```
`Error: 'coroutine' object has no attribute 'fetchall'`

Jan200101 replied:
I had this fixed already and accidentally reverted it :skull: 
fixed in 314d9d26f26e8eefe0c26da83a49b224881a7c81

Bobbyperson commented on cogs/economy.py:
```diff
@@ -2735,12 +2698,14 @@ async def leaderboard(self, ctx):
-                "SELECT user_id, balance FROM main ORDER BY CAST(balance AS INTEGER) DESC"
-            )
-            rows = await cursor.fetchall()
+        async with self.client.session as session:
```
Line 2712 still contains old code for aiosqlite, which causes an error due to it trying to subscript Main. You did the fix in `serverleaderboard`, I think you just forgot it here.

Jan200101 replied:
fixed in 6970eb7460028c2811c7cb25d9068af3dc4605a7

Bobbyperson commented on cogs/osu.py:
```diff
@@ -216,68 +192,81 @@ async def invest(self, ctx, osu_id: int = 0, amount: str | None = None):
+            results = (
+                await session.execute(
+                    select(models.osu.Osu).where(
+                        models.osu.Osu.osu_id == osu_id,
```
`AttributeError: type object 'Osu' has no attribute 'osu_id'`
I think you meant `osu_user`

Bobbyperson commented:
Thank you for your PR! Great changes overall, just needs some light edits. I did not test stocks.py since the stock market is closed (and I'm lazy), and I did not test fire.py since that sounds like a pain the ass.

Bobbyperson commented:
Did not mean to publish review yet, hit the wrong button. Some other general things are:
1. If a query does not return rows with the `sql` command, it returns `Error: This result object does not return rows. It has been closed automatically.`
2. When adding an unfire, the following traceback is given
```
Ignoring exception in on_raw_reaction_add
Traceback (most recent call last):
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/discord/client.py", line 481, in _run_event
    await coro(*args, **kwargs)
  File "/home/X/projects/nocaro/cogs/fire.py", line 263, in on_raw_reaction_add
    await self.add_to_board(added_msg, fireboard, total_reacts, "unfire")
  File "/home/X/projects/nocaro/cogs/fire.py", line 140, in add_to_board
    existing_message = await fireboard.fetch_message(exists)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/discord/abc.py", line 1710, in fetch_message
    data = await self._state.http.get_message(channel.id, id)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/discord/http.py", line 758, in request
    raise NotFound(response, data)
discord.errors.NotFound: 404 Not Found (error code: 10008): Unknown Message
```

Jan200101 commented:
fixed all the issues locally except unfire
![image](https://github.com/user-attachments/assets/a361c5bf-f432-4e40-a7fb-b903f3b34c0f)

Bobbyperson commented:
Now despite changing nothing I can't really get unfire to do anything at all (fire works fine). It's almost certainly my end so for now don't worry about it.

Bobbyperson commented on cogs/fire.py:
```diff
@@ -263,7 +241,7 @@ async def on_raw_reaction_add(self, payload):
                     break
         if react == fire:
-            if total_reacts >= 5:
+            if total_reacts >= 1:
```
This needs to be raised again when done testing

Jan200101 commented:
- squashed all changes
- moved from execution to scalars where its benefitial
- restored fire numbers
- replaced absolute bot imports with main imports
- added missing table entries for Fire
- turned some implicit transactions into explicit transactions

Bobbyperson commented on cogs/economy.py:
```diff
@@ -2755,14 +2727,14 @@ async def leaderboard(self, ctx):
+        for rank, main in enumerate(top_users.items(), start=1):
             try:
-                username = await self.client.fetch_user(user_id)
+                username = await self.client.fetch_user(main.user_ID)
```
`AttributeError: 'tuple' object has no attribute 'user_ID'`

Bobbyperson reviewed (changes requested):
the `highestfire` command gives this traceback (same issue is present with `highestunfire`)
```
Command raised an exception: AttributeError: Neither 'InstrumentedAttribute' object nor 'Comparator' object associated with Fire.guild_id has an attribute 'reacts'
Traceback (most recent call last):
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/sqlalchemy/sql/elements.py", line 1506, in __getattr__
    return getattr(self.comparator, key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Comparator' object has no attribute 'reacts'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/sqlalchemy/orm/attributes.py", line 472, in __getattr__
    return getattr(self.comparator, key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 1413, in __getattr__
    return self._fallback_getattr(key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/sqlalchemy/orm/properties.py", line 474, in _fallback_getattr
    return getattr(self.__clause_element__(), key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/sqlalchemy/sql/elements.py", line 1508, in __getattr__
    raise AttributeError(
AttributeError: Neither 'AnnotatedColumn' object nor 'Comparator' object has an attribute 'reacts'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/discord/ext/commands/core.py", line 235, in wrapped
    ret = await coro(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/X/projects/nocaro/cogs/fire.py", line 411, in highestfire
    .order_by(models.fire.Fire.guild_id.reacts.desc())
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/sqlalchemy/orm/attributes.py", line 474, in __getattr__
    raise AttributeError(
AttributeError: Neither 'InstrumentedAttribute' object nor 'Comparator' object associated with Fire.guild_id has an attribute 'reacts'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/discord/ext/commands/bot.py", line 1366, in invoke
    await ctx.command.invoke(ctx)
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/discord/ext/commands/core.py", line 1029, in invoke
    await injected(*ctx.args, **ctx.kwargs)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/X/projects/nocaro/.venv/lib/python3.12/site-packages/discord/ext/commands/core.py", line 244, in wrapped
    raise CommandInvokeError(exc) from exc
discord.ext.commands.errors.CommandInvokeError: Command raised an exception: AttributeError: Neither 'InstrumentedAttribute' object nor 'Comparator' object associated with Fire.guild_id has an attribute 'reacts'
```
Also, the osu related commands still seem to be broken
![image](https://github.com/user-attachments/assets/148bb439-3611-4cea-a518-2bb9e58357ea)
I think beyond `highest(un)fire`, fire.py is fully working.
After these couple of things are fixed, it'll be ready to merge. Thank you again for your PR!

Jan200101 commented:
stocks work again, unsure how the original code ever worked and the definition you shared doesn't explain it either.

Can you check if the current database has AUTO INCREMENT set? If so then this can be safely merged otherwise a workaround would need to be put in place until alembic is integrated (which is another headache)

Bobbyperson commented:
Current DB does not have auto increment, whoops. I can edit the DB schema to be auto increment, but I'll poke around beforehand to see what works the best.

Bobbyperson commented on cogs/stocks.py:
```diff
@@ -24,59 +23,42 @@ def __init__(self, client):
-                            (rowid,),
-                        )
-                        amount -= current_amount
+                    if row.current_amount <= amount:
```
`amount` not `current_amount`

Bobbyperson reviewed (changes requested):
Just this one tiny error and everything is good to merge!

Jan200101 commented:
> Current DB does not have auto increment, whoops.

some digging revealed that sqlite deals with this automatically
> On an [INSERT](https://sqlite.org/lang_insert.html), if the ROWID or INTEGER PRIMARY KEY column is not explicitly given a value, then it will be filled automatically with an unused integer, usually one more than the largest ROWID currently in use. This is true regardless of whether or not the AUTOINCREMENT keyword is used.

https://sqlite.org/autoinc.html

explicitly auto incrementing will mean there will never be an overlap of IDs, which I think is good and means no change to the database is needed.
Also sqlite does not support Altering Table Columns beyond the basics like name.

Bobbyperson commented:
I've altered pretty much all of the tables lots of times as I've needed with <https://sqlitebrowser.org/>. 

Regardless, thank you very very much for your PR, merging now.
